library/Vanilla/Formatting/DateTimeFormatter.php

```diff
@@ -0,0 +1,406 @@
+<?php
+/**
+ * @author Adam Charron <adam.c@vanillaforums.com>
+ * @copyright 2009-2019 Vanilla Forums Inc.
+ * @license GPL-2.0-only
+ */
+
+namespace Vanilla\Formatting;
+
+use Garden\StaticCacheTranslationTrait;
+
+/**
+ * Formatting methods related to dates & times.
+ */
+class DateTimeFormatter {
+
+    use StaticCacheTranslationTrait;
+
+    const FORCE_FULL_FORMAT = 'force-full-datetime-format';
+
+    /** @var DateConfig */
+    private $dateConfig;
+
+    /**
+     * @param DateConfig $dateConfig
+     */
+    public function __construct(DateConfig $dateConfig) {
+        $this->dateConfig = $dateConfig;
+    }
+
+    const NULL_TIMESTAMP_DEFALT_VALUE = '-';
+
+    /**
+     * Format a MySQL DateTime string in the specified format.
+     *
+     * @link http://us.php.net/manual/en/function.strftime.php
+     *
+     * @param string|number $timestamp A timestamp or string in Mysql DateTime format. ie. YYYY-MM-DD HH:MM:SS
+     * @param bool $isHtml Whether or not to output this as an HTML string.
+     * @param string $format The format string to use. Defaults to the application's default format.
+     * @return string
+     */
+    public function formatDate($timestamp = '', bool $isHtml = false, string $format = ''): string {
+        // Was a mysqldatetime passed?
+        if ($timestamp !== null && !is_numeric($timestamp)) {
+            $timestamp = self::dateTimeToTimeStamp($timestamp);
+        }
+
+        if ($timestamp === null) {
+            return self::t('Null Da' .
+                'te', self::NULL_TIMESTAMP_DEFALT_VALUE);
+        }
+
+        $gmTimestamp = $timestamp;
+        $timestamp = $this->adjustTimeStampForUser($timestamp);
+
+        if ($format == '') {
+            $format = $this->getDefaultFormatForTimestamp($timestamp);
+        } elseif ($format === self::FORCE_FULL_FORMAT) {
+            $format = $this->dateConfig->getDefaultDateTimeFormat();
+            $format = $this->normalizeFormatForTimeStamp($format, $timestamp);
+        }
+
+        $result = strftime($format, $timestamp);
+
+        if ($isHtml) {
+            $fullFormat = $this->dateConfig->getDefaultDateTimeFormat();
+            $fullFormat = $this->normalizeFormatForTimeStamp($fullFormat, $timestamp);
+            $result = wrap(
+                $result,
+                'time',
+                [
+                    'title' => strftime($fullFormat, $timestamp),
```

Comment:
Are we taking into account that `$fullFormat` is now partially rendered string but not just format pattern? (since `normalizeFormatForTimeStamp()` call)

Reply:
I believe it was. This is essentially a 1:1 translation of what was there before. It looks like it just preformats certain portions. 

Since `strftime()` is so flexible I don't see any issues there. If you can think of a test case I'd be glad to add it though.